### DISCUSSION TITLE
Some questions about writing custom functions.

### DISCUSSION THREAD

**Discussion by JinXiu2023** [open-policy-agent/community#458] [OPA and Rego] [answered]
Hello. I came across some questions when I wrote custom functions. I found that the custom function cannot use some keys like  `default` and `if`. I wonder if I want to write a custom function, what I can use in my function. I hope that the writing of custom function is like the writing of opa policy . Thank you very much.

The error example:
```
package example.test

import future.keywords.in
import future.keywords.if

test(params) := result{
    default defenseResult := "reject"
    defenseResult := "pass" if {
        count(params) <= 0
    }
    result := {
    	"result": defenseResult
    }
}
```
The error is :
```
1 error occurred: policy.rego:7: rego_parse_error: unexpected default keyword
	    default defenseResult := "reject"
	    ^
```

**Comment by srenatus** [comment, marked as answer]
Hello! I'm afraid you've got a few things slightly mixed up. You cannot define rules in functions (or in other rules), and that's what 
```rego
test(params) := result{
    default defenseResult := "reject"
    defenseResult := "pass" if {
        count(params) <= 0
    }
    result := {
    	"result": defenseResult
    }
}
```
looks like.

You can, however, achieve the desired behaviour differently:

```rego
package example.test

import future.keywords.if

test(params) := wrap("pass")   if count(params) == 0 # negative should be impossible
test(params) := wrap("reject") if count(params) > 0

wrap(res) := {"result": res}
```

**Reply by JinXiu2023** [reply]
Ok. Thank you for your answer. I'm going to split my error function into some little functions and try to avoid such bad uses.
